<!doctype html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <title>akt - Using akt</title>
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <!--[if lt IE 9]>
       <script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
       <script src="https://oss.maxcdn.com/libs/respond.js/1.4.2/respond.min.js"></script>
       <![endif]-->
  <link rel="stylesheet"
          href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.2/css/bootstrap.min.css">
  <link rel="stylesheet" type="text/css" href="css/screen.css">
  <link rel="stylesheet" type="text/css" media="print" href="css/print.css">
  <script src="https://ajax.googleapis.com/ajax/libs/jquery/1.11.1/jquery.min.js"
            type="text/javascript"></script>
  <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.2/js/bootstrap.min.js"></script>
  <script src="js/nav.js"></script>
  <script type="text/javascript">

  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-2234613-4']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();
  </script>
  <script type="text/javascript">
  /* <![CDATA[ */
    (function() {
        var s = document.createElement('script'), t = document.getElementsByTagName('script')[0];
        s.type = 'text/javascript';
        s.async = true;
        s.src = 'http://api.flattr.com/js/0.6/load.js?mode=auto';
        t.parentNode.insertBefore(s, t);
    })();
  /* ]]> */
  </script>
</head>
<body>
  <div id="doc" class="container-fluid">
    <header>
    <header>
      <div id="flattr">
 	<a href="https://github.com/illumina/akt">github.com/illumina/akt</a>
      </div>

      <!-- <div id="github"> -->
      <!-- 	<a href="https://github.com/illumina/akt">github.com/illumina/akt</a> -->
      <!-- </div> -->

        <span class="big">akt</span>
        &nbsp;
        <span class="small">ancestry and kinship tools</span>
    </header>
    <div id="bd">
      <nav id="navbar" class="navbar nav navbar-default col-md-3 col-sm-4">
        <div class="navbar-header">
  <button type="button" class="navbar-toggle"
          data-toggle="collapse"
          data-target=".navbar-collapse">
    <span class="sr-only">Toggle navigation</span>
    <span class="icon-bar"></span>
    <span class="icon-bar"></span>
    <span class="icon-bar"></span>
  </button>
</div>
<div class="navbar-collapse collapse">
  <ul class="nav tree">
    <li><a href="index.html">About</a></li>
    <li><a href="license.html">License</a></li>
    <li><a href="usage.html">Usage</a></li>
    <li><a href="examples.html">Examples</a></li>
    <li><a href="help.html">Help</a></li>
  </ul>
</div>
        <div id="toc">
<ul>
<li><a href="#pca">pca</a></li>
<li><a href="#kin">kin</a></li>
<li><a href="#relatives">relatives</a></li>
<li><a href="#mendel">mendel</a></li>
<li><a href="#metafreq">metafreq</a></li>
<li><a href="#cluster">cluster</a></li>
<li><a href="#stats">stats</a></li>
<li><a href="#ibd">ibd</a></li>
<li><a href="#ldplot">LDplot</a></li>
<li><a href="#admix">admix</a></li>
</ul>
        </div>
      </nav>
      <main class="col-md-9 col-sm-8">
        <div class="row">
          <h1 class="title">Using akt</h1>
<p>akt uses the syntax</p>
<pre><code>./akt COMMAND &lt;options&gt;</code></pre>
<p>To see a list of available tools use</p>
<pre><code>
Program:    akt (Ancestry and Kinship Tools)
Version:    da6bc6d
Copyright (c) 2016, Illumina, Inc. All rights reserved. See LICENSE for further details.

Usage:  akt &lt;command&gt; [options]

    pca                      principal component analysis
    kin                      detect average IBD sharing
    relatives                discover pedigrees
    ibd                      detect segments shared IBD
    mendel                   profile Mendelian inhertiance and inconsistencies in known pedigrees
    cluster                  perform cluster analyses
    LDplot                   output correlation matrix
    stats                    calculate AF and LD metrics
    metafreq                 examine two files for AF differences
</code></pre>
<h2 id="pca"><a href="#pca">pca</a></h2>
<pre><code>Performs principal component analysis on a vcf/bcf
Usage:
./akt pca input.bcf
     -T --targets-file:     similar to -R but streams rather than index-jumps
     -o --output:           output vcf
     -O --outputfmt:        output vcf format
     -r --regions:          chromosome region
     -R --regions-file:     restrict to regions listed in a file
     -s --samples:          list of samples
     -S --samples-file:     list of samples, file
     -h --thin:         keep every h variants
     -m --maf:          minimum MAF
     -w --weight:           VCF with weights for PCA
     -N --npca:         first N principle components
     -a --alg:          exact SVD (slow)
     -C --covdef:           definition of SVD matrix: 0=(G-mu) 1=(G-mu)/sqrt(p(1-p)) 2=diag-G(2-G) default(1)
     -e --extra:            extra vectors for Red SVD
     -F --svfile:           File containing singular values
</code></pre>
<ul>
<li>-T : Indexed VCF file containing intersecting sites and relevant site info.</li>
<li>-o : Output to vcf.</li>
<li>-O : Output format of vcf b=compressed bcf, z=compressed vcf.</li>
<li>-r : Comma-separated list of regions, chr:from-to.</li>
<li>-R : File containing 3 columns: CHROM, POS and POS_TO.</li>
<li>-s : Comma-separated list of samples to include or exclude if prefixed with “^”.</li>
<li>-S : File of sample names to include or exclude if prefixed with “^”. One sample per line.</li>
<li>-h : Hop this many markers forward from last accepted marker.</li>
<li>-m : Minimum MAF a site to be counted. When reading: if AF &gt; 0.5 checks 1-AF &gt; m.</li>
<li>-w : Use precalculated principle components.</li>
<li>-N : Number of principle components to calculate.</li>
<li>-a : Use JacobiSVD PCA algorithm, which is exact to float precision but very slow.</li>
<li>-e : Default PCA calculation is the inexact <code>RedSVD</code> algorithm, which requires this parameter. The higher the number the more accurate principle components will be obtained.</li>
<li>-F : File to output the singular values.</li>
<li>-C : Which matrix to take the PCA of. 0 uses mean subtracted genotype matrix; 1 uses mean subtracted and normalized genotype matrix; 2 uses normalized covariance matrix with bias term subtracted from diagonal elements.</li>
</ul>
<p><code>pca</code> can either calculate principle components from data and do projections onto these components or use precalculated principle component vectors</p>
<pre><code>~/ancestry_tools$ ./akt pca multisample.bcf -O b -o pca.bcf -m 0.05 -h 10 -N 5 &gt; projections </code></pre>
<p>The file <code>projections</code> contains</p>
<pre><code>SAMPLE_ID0 P0 P1 P2 P3 P4
SAMPLE_ID1 P0 P1 P2 P3 P4
...</code></pre>
<p>The bcf file <code>pca.bcf</code> contains</p>
<pre><code>bcftools query -f &quot;%INFO/WEIGHT\n&quot; pca.bcf
pc00 pc01 pc02 pc03 pc04
pc10 pc11 pc12 pc13 pc14
...</code></pre>
<p>First index is the site index and second is which principle component. If you already have precalculated PCA weights and you want to project new samples onto those weights</p>
<pre><code>./akt pca new_multisample.bcf -w pca.bcf &gt; projections</code></pre>
<h2 id="kin"><a href="#kin">kin</a></h2>
<pre><code>
About: Calculate kinship/IBD statistics from a multisample BCF/VCF
Usage: akt kin [options] &lt;in.bcf&gt;
Expects input.bcf to contain genotypes.

Kinship calculation options:
     -k --minkin:           threshold for relatedness output (none)
     -F --freq-file:                a file containing population allele frequencies to use in kinship calculation
     -M --method:           type of estimator. 0:plink (default) 1:king-robust 2:genetic-relationship-matrix
     -a --aftag:            allele frequency tag (default AF)
     -n --nthreads:         num threads

Site filtering options:
     -R --regions-file:     restrict to regions listed in a file
     -r --regions:          chromosome region
     -T --targets-file:     similar to -R but streams rather than index-jumps
     -t --targets:              similar to -r but streams rather than index-jumps
     -m --maf:          minimum MAF
     -h --thin:         keep every h variants
        --force:            run kin without -R/-T/-F

Sample filtering options:
     -s --samples:          list of samples
     -S --samples-file:     list of samples, file

</code></pre>
<ul>
<li>-k : Only output pairs with kinship coefficient greater than this.</li>
<li>-F : a file containing population allele frequencies to use in kinship calculation</li>
<li>-M : type of estimator. 0:<a href="https://www.cog-genomics.org/plink2/ibd">plink (default)</a> 1:<a href="http://bioinformatics.oxfordjournals.org/content/26/22/2867.full">king-robust</a> 2:<a href="http://cnsgenomics.com/software/gcta/estimate_grm.html">genetic-relationship-matrix</a></li>
<li>-a : Allele frequency tag e.g. “TAG” reads or writes “TAG_AF” in the target VCF.</li>
<li>-n : Number of threads to use.</li>
<li>-R : File containing 3 columns: CHROM, POS and POS_TO.</li>
<li>-r : Comma-separated list of regions, chr:from-to.</li>
<li>-T : Indexed VCF file containing intersecting sites and relevant site info.</li>
<li>-t : similar to -r but streams rather than index-jumps.</li>
<li>-m : Minimum MAF a site to be counted. When reading: if AF &gt; 0.5 checks 1-AF &gt; m.</li>
<li>-h : Hop this many markers forward from last accepted marker.</li>
<li>-s : Comma-separated list of samples to include or exclude if prefixed with “^”.</li>
<li>-S : File of sample names to include or exclude if prefixed with “^”. One sample per line.</li>
</ul>
<p>Run the kinship calculation by giving akt a multi-sample vcf/bcf file:</p>
<pre><code>$ akt kin multisample.bcf -R data/wgs.grch37.vcf.gz -n 32 &gt; kin.txt</code></pre>
<h3 id="choice-of-estimator"><a href="#choice-of-estimator">Choice of estimator</a></h3>
<p>The default algorithm (<code>-M 0</code>) used to calculate IBD is taken from <a href="http://www.ncbi.nlm.nih.gov/pmc/articles/PMC1950838/">PLINK</a> with some minor changes. It outputs the following seven column format:</p>
<pre><code>ID1 ID2 IBD0 IBD1 IBD2 KINSHIP NSNP</code></pre>
<p>As with PLINK, we set KINSHIP = 0.5 * IBD2 + 0.25 * IBD1. Our IBD values may slighly differ to PLINK’s (by desing) due to the following differences:</p>
<ul>
<li>No ‘bias correction’ since allele frequencies are assumed to be accurate</li>
<li>Normalization as follows:
<ul>
<li>if IBD0 &gt; 1: IBD0 = 1, IBD1 = 0, IBD2 = 0</li>
<li>if IBD1 &lt; 0: IBD1 = 0</li>
<li>if IBD2 &lt; 0: IBD2 = 0</li>
<li>norm = IBD0 + IBD1 + IBD2</li>
<li>IBD0 /= norm, IBD1 /= norm; IBD2 /= norm;</li>
</ul></li>
<li>We do <strong>not</strong> follow PLINK which forces IBD to obey consistency conditions - this affects the clustering that is required for the <code>relatives</code> code.</li>
</ul>
<p>The second method (<code>-M 1</code>) uses the robust kinship coefficent estimate describing in the <a href="http://bioinformatics.oxfordjournals.org/content/26/22/2867.full">KING paper</a>. This may be preferable when your cohort has large amounts of population structure. The IBD estimates and output format are the same as for <code>-M 0</code>.</p>
<p>The third method (<code>-M 2</code>) outputs the genetic relationship matrix used by <a href="http://cnsgenomics.com/software/gcta/">GCTA</a> (among other software). The output format is <code>ID1 ID2 GR</code> where <code>GR</code> is the genetic correlation between samples <code>ID1</code> and <code>ID2</code>. This matrix is important for fitting linear mixed-effect models but we have found methods 0 and 1 more appropriate for sample QC purposes.</p>
<h2 id="relatives"><a href="#relatives">relatives</a></h2>
<pre><code>Discover relatives from IBD
Usage:
./akt relatives ibdfile
     -k --kmin:         threshold for relatedness (0.05)
     -i --its:          number of iterations to find unrelated (10)
     -g --graphout:         if present output pedigree graph files
     -p --prefix:           output file prefix (out)
arrow types     : solid black   = parent-child
                : dotted black  = siblings
                : blue      = second order
                : red       = duplicates
                : directed  = from parent to child
</code></pre>
<ul>
<li>-k : Only keep links with kinship above this threshold (searches in this set for duplicate, parent-child and sibling links).</li>
<li>-i : Iteration parameter for unrelated set output.</li>
<li>-p : Prefix for output files.</li>
<li><p>-g : If present output graphviz files. These can be visulaised using e.g. <code>neato -Tpng -O out.allgraph</code> or for family pedigrees <code>dot -Tpng -O out.Fam0.graph</code>.</p>
<pre><code>./akt relatives allibd -g &gt; allrelatives</code></pre>
<p>The output contains duplicates, families, relationship types and unrelated individuals</p>
<pre><code>grep ^Dup allrelatives
Dup0 Sample0
Dup0 Sample1
...
grep ^Fam allrelatives
Fam0 Sample2
Fam0 Sample3
...
...
grep ^Type allrelatives
Type Fam0 Sample2 Sample3 Parent/Child
...
grep ^Unrel allrelatives
Sample0
Sample2
...</code></pre>
<p>The file <code>out.allgraph</code> can be viewed with gviz e.g. <code>fdp out.allgraph -Tpng -O</code> and the families can be viewed using e.g. <code>dot out.Fam0.graph -Tpng -O</code>. The parent child relations are also recorded in PLINK fam format in <code>out.fam</code>. If e.g. a sibling pair, is found the samples will appear in <code>out.fam</code> without parents. If the direction of the relationship can’t be determined e.g. for parent/child duos a random sample is assigned to be the parent in <code>out.fam</code>. The final column in the <code>.fam</code> file specifies how many potential parents the sample had.</p></li>
</ul>
<h2 id="mendel"><a href="#mendel">mendel</a></h2>
<pre><code>
About:   akt mendel - profiles duo/trios
Usage:   ./akt mendel input.bcf -p pedigree.fam

Options:
    -p, --pedigree              pedigree information in plink .fam format
    -o, --out                   output a site only vcf.gz annotated with site specific error rates
    -i, --include               variant filters to apply eg. -i &#39;TYPE==snp &amp;&amp; QUAL&gt;=10 &amp;&amp; DP&lt;100000 &amp;&amp; HWE&lt;10&#39; 
    -t, --targets [^]&lt;region&gt;   Set regions. Exclude regions with &quot;^&quot; prefix
    -r, --regions &lt;region&gt;      restrict to comma-separated list of regions
    -R, --regions-file &lt;file&gt;   restrict to regions listed in a file</code></pre>
<ul>
<li>-o : Site only vcf annotated with mendel error rates</li>
<li>-i : Apply filters see <a href="https://samtools.github.io/bcftools/bcftools.html#expressions">bcftools expressions</a></li>
<li>-t : Regions/sites to include/exclude see <a href="https://samtools.github.io/bcftools/bcftools.html#common_options">bcftools</a></li>
<li>-r : Comma-separated list of regions, chr:from-to.</li>
<li>-R : File containing 3 columns: CHROM, POS and POS_TO.</li>
<li>-p : The <code>-p</code> option specifies the pedigree structure and uses the well-known six column plink .fam format:</li>
</ul>
<ol type="1">
<li>Family ID</li>
<li>Individual ID</li>
<li>Paternal ID</li>
<li>Maternal ID</li>
<li>Sex (1=male; 2=female; other=unknown)</li>
<li>Phenotype Here is an example pedigree file for the <a href="http://www.nature.com/nmeth/journal/v12/n10/fig_tab/nmeth.3505_SF4.html">CEPH1463</a> pedigree:</li>
</ol>
<pre><code>PG NA12889 0 0  1 -9
PG NA12890 0 0  2 -9
PG NA12891 0 0  1 -9
PG NA12892 0 0  2 -9
PG NA12877 NA12889 NA12890      1 -9
PG NA12878 NA12891 NA12892      2 -9
PG NA12879 NA12877 NA12878      2 -9
PG NA12880 NA12877 NA12878      2 -9
PG NA12881 NA12877 NA12878      2 -9
PG NA12882 NA12877 NA12878      1 -9
PG NA12883 NA12877 NA12878      1 -9
PG NA12884 NA12877 NA12878      1 -9
PG NA12885 NA12877 NA12878      2 -9
PG NA12886 NA12877 NA12878      1 -9
PG NA12887 NA12877 NA12878      2 -9
PG NA12888 NA12877 NA12878      1 -9
PG NA12893 NA12877 NA12878      1 -9</code></pre>
<h2 id="metafreq"><a href="#metafreq">metafreq</a></h2>
<pre><code>Compare AFs between two cohorts
Usage:
akt metafreq a.vcf.gz b.vcf.gz -Oz -o meta.sites.vcf.gz
     -r --regions:          chromosome region
     -R --regions-file:     restrict to regions listed in a file
     -T --targets-file:     similar to -R but streams rather than index-jumps
     -o --output:           output vcf
     -O --outputfmt:        output vcf format
     -a --aftag:            allele frequency tag (default AF)</code></pre>
<ul>
<li>-T : Indexed VCF file containing intersecting sites and relevant site info.</li>
<li>-r : Comma-separated list of regions, chr:from-to.</li>
<li>-R : File containing 3 columns: CHROM, POS and POS_TO.</li>
<li>-a : Allele frequency tag e.g. “TAG” reads or writes “TAG_AF” in the target VCF.</li>
<li>-o : Output to vcf.</li>
<li>-O : Output format of vcf b=compressed bcf, z=compressed vcf.</li>
</ul>
<p>This tool lets us find sites with statistically significant differences in frequency between two samples.</p>
<pre><code>./akt ibd f1.bcf -p f2.bcf -O b -o out.bcf </code></pre>
<p>The file <code>out.bcf</code> contains two fields <code>QF</code> and <code>QX</code> equal to -log10 of the p-value calculated using Fisher’s exact test and the chi-squared test, respectively. Higher scores indicate a more statistically significant difference. These values are capped at 100 and typically the chi-squared test is more likely to reject the null-hypothesis that the two cohorts have identical distributions.</p>
<h2 id="cluster"><a href="#cluster">cluster</a></h2>
<pre><code>Clustering on text files
Usage:   ./akt cluster input.txt
     -k --K:            number of clusters
     -i --seed:         random seed for starting values
     -a --alg:          clustering algorithm 0 = k++means, 1 = gaussian mixture, 2 = density method
     -c --cols:         column range to read
     -C --cfile:            initial guess for cluster centres in text file
     -o --outputcfile:      assigned cluster centres
     -c --cols:         which columns to use e.g. 2-4
     -I --maxits:           max number of iterations to use: a=0,1
     -d --dc:           radius for density method: a=2
     -p --rho_min:          min density for cluster centre: a=2
     -D --delta_min:        min radius for cluster centre: a=2
     --density-plot:        plot the density and finish: a=2
     -e --silhouette:       calculate silhouette score</code></pre>
<ul>
<li>-k : Number of clusters. Examine the data to guess this or analyse silhouette scores.</li>
<li>-i : Random seed.</li>
<li>-a : 0 to use k++ means clustering, 1 to use EM with Gaussians, 2 to use density method.</li>
<li>-C : file with initial guess for cluster centres. If analysing 2d data this should contain 2 columns and K rows.</li>
<li>-c : Which columns in input file to use.</li>
<li>-I : Maximum number of iterations to use for alg 0 or 1.</li>
<li>-d : radius around each point for counting density.</li>
<li>-p : min density for cluster centre.</li>
<li>-D : min radius for cluster centre.</li>
<li>–density-plot : plot delta-density graph and finish.</li>
<li>-e : Calculate silhouette score (goodness of cluster assignment).</li>
</ul>
<p>e.g. cluster the first three principle components from the output of vcfpca using k++ means. Different clusters should correspond to different ancestries</p>
<pre><code>./akt cluster projections -k 4 -c 2-4 -e &gt; clustered</code></pre>
<p>Each block in <code>clustered</code> contains all the data for points in that cluster with the row layout</p>
<pre><code>P2 P3 P4 (SILHOUETTE) CLUSTERID P1 P5 ...</code></pre>
<p>To visually inspect the clustered data try the following gnuplot command</p>
<pre><code>gnuplot&gt; splot for [i=0:3:1] &#39;clustered&#39; every:::i::i</code></pre>
<p>The silhouette column can be used to filter on badly clustered samples. ~1 means a sample is closer to the centroid of its assigned cluster than the next nearest cluster. A sample can be badly clustered for many reasons: bad choice of K, convergence of the clustering algorithm to bad local minimum, individual is of mixed ancestry. You could filter out badly clustered samples using</p>
<pre><code>awk &#39;{ if($4 &gt; 0.5) print $0}&#39; clustered &gt; well_clustered</code></pre>
<h3 id="density-clustering"><a href="#density-clustering">density clustering</a></h3>
<p>This is a better clustering algorithm but requires some tuning by hand. First run</p>
<pre><code>./akt cluster projections -c 2-4 -a 2 -d 1 --density-plot &gt; dplot</code></pre>
<p>the file dplot is a two column data file of density versus delta (the distance to the closest data point of higher density). By examining this plot outliers are obvious and a cutoff in density and delta can be determined that isolates these points. Once this cutoff is determined run</p>
<pre><code>./akt cluster projections -c 2-4 -a 2 -d 1 -p 0 -D 75 &gt; clustered</code></pre>
<p>Where 0 and 75 are the cutoffs determined from <code>dplot</code>. This will output the number of clusters that were found (including one for unassigned points) and <code>clustered</code> will contain the clustered data in the usual format.</p>
<pre><code>gnuplot&gt; splot for [i=0:4:1] &#39;clustered&#39; every:::i::i</code></pre>
<p>Unassigned data points are always put in cluster 0. The silhouette score for unassigned data is undefined.</p>
<h2 id="stats"><a href="#stats">stats</a></h2>
<pre><code>Calculate correlation across a population
Usage:
./akt stats input_filename.vcf
Expects input_filename.vcf to contain hard genotypes
     -F --flank:            size of left and right flanking region (1000bp)
     -b --block:            -F argument is number of markers instead of number of base pairs
     -x --afonly:           calculate allele freq only
     -c --output_cor:       output sitewise correlation (false)
     -C --output_cormin:        output sitewise correlation greater than (0)
     -a --aftag:            allele frequency tag (default AF)
     -o --output:           output vcf
     -O --outputfmt:        output vcf format
     -r --regions:          chromosome region
     -R --regions-file:     restrict to regions listed in a file
     -s --samples:          list of samples
     -S --samples-file:     list of samples, file</code></pre>
<ul>
<li>-F : Correlation with variants in a window of size f base pairs to left and right or each variant.</li>
<li>-b : The number in the -F argument now interpreted as number of flanking variants instead of flanking positions.</li>
<li>-x : Do not calculate covariance matrix.</li>
<li>-c : Output all correlation.</li>
<li>-C : Output correlation values greater than this.</li>
<li>-a : Allele frequency tag e.g. “TAG” reads or writes “TAG_AF” in the target VCF.</li>
<li>-o : Output to vcf.</li>
<li>-O : Output format of vcf b=compressed bcf, z=compressed vcf.</li>
<li>-r : Comma-separated list of regions, chr:from-to.</li>
<li>-R : File containing 3 columns: CHROM, POS and POS_TO.</li>
<li>-s : Comma-separated list of samples to include or exclude if prefixed with “^”.</li>
<li>-S : File of sample names to include or exclude if prefixed with “^”. One sample per line.</li>
</ul>
<p>This tool lets us calculate allele frequencies and correlation matrices from multisample vcfs. Running</p>
<pre><code>./akt stats panel.bcf -F 10000 -O b -o tmp.bcf -a &quot;TAG&quot; -c</code></pre>
<p>The output, tmp.bcf, has no sample columns and info fields * TAG_AF : Allele frequencies. * TAG_SIG : Allele variances. * TAG_LD : LDscore = sum of COR. * TAG_COR : correlation of this site with neighbouring sites. * TAG_CORP : position of neighbouring sites.</p>
<p>Correlation is bounded by [-1,1]. With the <code>-f 10000</code> option it will output the correlation of each site with the sites closer than 10000bp to the left and right. The LD metric is described in <a href="http://www.nature.com/ng/journal/v47/n3/full/ng.3211.html">this paper</a></p>
<h2 id="ibd"><a href="#ibd">ibd</a></h2>
<pre><code>Find IBD regions from a VCF
Usage:
./akt ibd in.bcf -p sites.bcf
Expects input_filename.vcf to contain hard genotypes
     -T --targets-file:     similar to -R but streams rather than index-jumps
     -r --regions:          chromosome region
     -R --regions-file:     restrict to regions listed in a file
     -s --samples:          list of samples
     -S --samples-file:     list of samples, file
     -h --thin:         keep every h variants
     -a --aftag:            allele frequency tag (default AF)
     -f --pairfile:         file containing sample pairs to perform calculations on
     -n --nthreads:         num threads
     -m --maf:          minimum MAF
     -e --error:            If no GQ then this is error probability(1e-3)
     -M --thresh:           likelihood output threshold (default 50)
     -L --length:           length output threshold (default 100000)
     -w --wsize:            window size (default 20)
     -x --maxerr:           stop when encountering a window with this many ( 0/0 , 1/1 ) pairs (default 2)
     -l --lsize:            try to join long regions closer than this (default 100000)</code></pre>
<ul>
<li>-T : Indexed VCF file containing intersecting sites and relevant site info.</li>
<li>-r : Comma-separated list of regions, chr:from-to.</li>
<li>-s : Comma-separated list of samples to include or exclude if prefixed with “^”.</li>
<li>-S : File of sample names to include or exclude if prefixed with “^”. One sample per line.</li>
<li>-h : Hop this many markers forward from last accepted marker.</li>
<li>-a : Allele frequency tag e.g. “TAG” reads or writes “TAG_AF” in the target VCF.</li>
<li>-f : File containing pairs of samples to compute, 2 column: Sample1 Sample2.</li>
<li>-n : Number of threads to use.</li>
<li>-m : Minimum MAF a site to be counted. When reading: if AF &gt; 0.5 checks 1-AF &gt; m.</li>
<li>-e : Probability of error per base</li>
<li>-M : Lower bound of likelihood ratio score</li>
<li>-L : Lower bound of segment length</li>
<li>-w : Number of markers per window</li>
<li>-x : Number of hom-ref hom-alt matches allowed in a window before breaking a region in first stage.</li>
<li>-l : Minimum IBD region to output</li>
</ul>
<p>This tool lets us find regions which are shared IBD between two samples.</p>
<pre><code>./akt ibd input.bcf -R panel.bcf -r 2 &gt; sharing</code></pre>
<p>The file <code>sharing</code> is in the format</p>
<pre><code>SAMPLE1 SAMPLE2 START END SCORE MATCHES LEN</code></pre>
<p>Where higher scores indicate a segment that is more likely to be shared IBD between samples 1 and 2. Matches are the number of IBS consistent matches between the two segments and LEN=END-START is the length of the segment.</p>
<p>The algorithm is a variation of <a href="https://data.broadinstitute.org/alkesgroup/Eagle/">EAGLE</a>. First we scan along the genome in windows of <code>w</code> counting the number of opposite homozygotes. With no errors an opposite homozygote indicates that the two windows cannot be IBD. Once we have counted in all the windows we join up neighbouring windows to create seed regions and attempt to extend these regions through windows with (w-x-1) opposite homozygotes. We calculate a running score based on the method in <a href="http://biorxiv.org/content/early/2015/12/18/028282">this paper</a>. Once these segments are calculated we try to join nearby segments (closer than <code>l</code>) if the score of the combined segment is greater than the score of either one separately. By filtering on segment length and score this code does a reasonable job of identifying IBD shared segments longer than about 1Mb from unphased data.</p>
<h2 id="ldplot"><a href="#ldplot">LDplot</a></h2>
<pre><code>Calculate correlation across a population
Usage:
./akt ldplot input_filename.vcf
Expects input_filename.vcf to contain hard genotypes
     -r --regions:          chromosome region
     -R --regions-file:     restrict to regions listed in a file
     -s --samples:          list of samples
     -S --samples-file:     list of samples, file</code></pre>
<ul>
<li>-r : Comma-separated list of regions, chr:from-to.</li>
<li>-R : File containing 3 columns: CHROM, POS and POS_TO.</li>
<li>-s : Comma-separated list of samples to include or exclude if prefixed with “^”.</li>
<li><p>-S : File of sample names to include or exclude if prefixed with “^”. One sample per line. This generates the correlation matrix of all included variants.</p>
<pre><code>./akt LDplot input.bcf -r 20:0-200000 &gt; sigma</code></pre>
<p>the file <code>sigma</code> contains the correlation matrix of all variants in the region <code>20:0-200000</code> and can be plotted with various tools e.g.</p>
<pre><code>gnuplot&gt; plot &#39;sigma&#39; matrix with image</code></pre></li>
</ul>
<h2 id="admix"><a href="#admix">admix</a></h2>
<pre><code>Approximate admixture fractions
Usage:   ./akt admix input.txt -C centres -c 2-3 
     -c --cols:         column range to read
     -C --cfile:            initial guess for cluster centres in text file
</code></pre>
<ul>
<li>-c : Which columns in input file to use.</li>
<li>-C : file with vectors whose transformation to admixture fractions is known.</li>
</ul>
<p><code>admix</code> attempts to use the output of <code>pca</code> to assign admixture fractions to data based on known populations. The input file should be in the form</p>
<pre><code>e1 e2 e3 ... ed POP a1 a2 a3 ... ad
...</code></pre>
<p>where <code>e1...</code> are coordinates in the space of principal components and <code>a1...</code> are coordinates in the space of admixture fractions. The mapping from PCA space to admixture space is</p>
<pre><code>Ta + v = e</code></pre>
<p>with inverse</p>
<pre><code>a = T^{-1}(e-v)</code></pre>
<p><code>admix</code> calculates this inverse mapping.</p>
<pre><code>~/ancestry_tools$ ./akt admix pcadata -c 2-6 -C pca_to_admix.txt &gt; admixtures</code></pre>
<p>The file <code>admixtures</code> contains</p>
<pre><code>SAMPLE_ID %POP1 %POP2 ... %POPD
...</code></pre>
<p>A vector must be chosen to fix the zero point of the transformation (the vector <code>v</code> above). For K populations the <code>-C</code> input file can contain K or K+1 lines. If K+1 vectors are given the one with the lowest norm in the space of admixtures equals <code>v</code>. If K vectors and their images in admixture space are specified then the zero point is fixed by choosing an arbitrary vector orthogonal to the sides of the tetrahedron made by the K inputs and roughly equidistant from them. There are two possible choices of sign: the positive one is always chosen.</p>
        </div>
      </main>
    </div>
    <footer>
    </footer>
  </div>
</body>
</html>
